### PR TITLE
[Material UI] Stack Amount / Required Amount

### PR DESCRIPTION
**Pull request by Kamesuta**
I love Schematica very much.
but **Material UI** is difficult for me to use.
I'd like to easier to collect materials.

I have **two changes** to make it more useful.

1. Add _Stack Amount_.

2. Add _Required Amount_.

I hope Schematica will be more wonderful mod!

### CODE REVIEW

**Comment by Lunatrius** [comment]
Took me a bit longer than usual to get to this. The available/missing items thing does sound like a neat feature but requires some tweaking in places, particularly for localization and item stack sizes.

I'm not entirely sure about the stack sizes as they're only really useful for smaller schematics and are pretty much useless for bigger schematics.

The import reordering is just adding noise and will be reverted the next time I run the formatter on those classes. Similarly the one line comments on all the changed lines don't add anything but noise.

File: src/main/java/com/github/lunatrius/schematica/client/gui/control/GuiSchematicMaterialsSlot.java [hunk 1]
```diff
@@ -44,6 +49,22 @@ protected void drawBackground() {
     protected void drawContainerBackground(Tessellator tessellator) {
     }
 
+    /**
+     * Scan your inventory item counts
+     */
+    protected void updateCount() {
```
Comment (Lunatrius) [inline src/main/java/com/github/lunatrius/schematica/client/gui/control/GuiSchematicMaterialsSlot.java]: `updateInventoryAmount` is less ambiguous. I'll have to check the 1.7.10 version, but there should already be a method in LunatriusCore that counts inventory items.

File: src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java [hunk 2]
```diff
@@ -82,6 +83,7 @@ private WrappedItemStack findOrCreateWrappedItemStackFor(final List<WrappedItemS
         public ItemStack itemStack;
         public int placed;
         public int total;
+        public int hold; // cache current hold item counts
```
Comment (Lunatrius) [inline src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java]: `inventory` is more fitting, `hold` implies it uses only the "active" item stack.

File: src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java [hunk 3]
```diff
@@ -98,8 +100,34 @@ public String getItemStackDisplayName() {
...
-            return String.format("\u00a7%c%d\u00a7r/%d", color, this.placed, this.total);
+            // Add stack count to easier to collect items! "counts(stack:remainder)"
+            return String.format("\u00a7%c%d(%d:%d)\u00a7r/%d(%d:%d)", color, placed, placedstack, placedremainder, total, totalstack, totalremainder);
+        }
+
+        /**
+         * Show you How much items should you collect.
+         * @return require item counts
+         */
+        public String getFormattedCount() {
```
Comment (Lunatrius) [inline src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java]: The name is pretty much the same as `getFormattedAmount`. `getFormattedAmountRequired` might be a better and more descriptive name.

The math/logic part also seems to be off as well, it doesn't take into account placed blocks at all.

File: src/main/java/com/github/lunatrius/schematica/client/gui/control/GuiSchematicMaterialsSlot.java [hunk 4]
```diff
@@ -19,6 +22,8 @@ public GuiSchematicMaterialsSlot(GuiSchematicMaterials par1) {
         super(Minecraft.getMinecraft(), par1.width, par1.height, 16, par1.height - 34, 24);
         this.guiSchematicMaterials = par1;
         this.selectedIndex = -1;
+        // run scanning
+        this.updateCount();
```
Comment (Lunatrius) [inline src/main/java/com/github/lunatrius/schematica/client/gui/control/GuiSchematicMaterialsSlot.java]: A nitpick, but I usually don't prefix methods with `this`. *They are unique in any context whereas variables might be either a field, local variable, parameter, ... in the same context.*

**Comment by Kamesuta** [comment]
You pointed out is very sharp, it has become a reference for me.
Problems and of **the stack size**, the **name** of the method, **localization**, etc., has solved all of the problems that gave us pointed out.
Statement that the import statement is formatted by the eclipse was returned to the original.
I am praying that your mod the better.

File: src/main/java/com/github/lunatrius/schematica/client/gui/control/GuiSchematicMaterialsSlot.java [hunk 5]
```diff
@@ -46,22 +46,6 @@ protected void drawBackground() {
     protected void drawContainerBackground(Tessellator tessellator) {
     }
 
-    /**
-     * Scan your inventory item counts
-     */
-    protected void updateCount() {
```
Comment (Kamesuta) [inline src/main/java/com/github/lunatrius/schematica/client/gui/control/GuiSchematicMaterialsSlot.java]: Use LunatriusCore's method.
「com.github.lunatrius.core.entity.EntityHelper.getItemCountInInventory(IInventory inventory, Item item, int itemDamage)」

File: src/main/java/com/github/lunatrius/schematica/client/gui/control/GuiSchematicMaterialsSlot.java [hunk 6]
```diff
@@ -13,6 +16,9 @@
 
     private final GuiSchematicMaterials guiSchematicMaterials;
 
+    private final String strMaterialRequired = I18n.format(Names.Gui.Control.MATERIAL_REQUIRED);
```
Comment (Kamesuta) [inline src/main/java/com/github/lunatrius/schematica/client/gui/control/GuiSchematicMaterialsSlot.java]: preparing for gui localization

File: src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java [hunk 7]
```diff
@@ -63,6 +64,13 @@
             }
         }
 
+        for (WrappedItemStack wrappedItemStack : blockList) {
```
Comment (Kamesuta) [inline src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java]: Update item amount count

File: src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java [hunk 8]
```diff
@@ -63,6 +64,13 @@
...
+            if (player.capabilities.isCreativeMode)
+                wrappedItemStack.inventory = -1;
```
Comment (Kamesuta) [inline src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java]: If the player is creative mode, it always shows "Available".

File: src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java [hunk 9]
```diff
@@ -99,7 +108,27 @@ public String getItemStackDisplayName() {
...
+        public String getFormattedAmountRequired(final String reqstr, final String avastr) {
+            final int need = this.total - this.inventory - this.placed;
+            if (this.inventory != -1 && need > 0) {
+                return String.format("\u00a7c%s:%s", reqstr, getFormattedStackAmount(itemStack, need));
+            } else {
+                return String.format("\u00a7a%s", avastr);
+            }
+        }
+
+        private static String getFormattedStackAmount(final ItemStack itemStack, final int amount) {
```
Comment (Kamesuta) [inline src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java]: Only amount being more than stack size, it shows stack size and its remainder

File: src/main/java/com/github/lunatrius/schematica/reference/Names.java [hunk 10]
```diff
@@ -160,6 +160,8 @@
 
             public static final String MATERIAL_NAME = "schematica.gui.materialname";
             public static final String MATERIAL_AMOUNT = "schematica.gui.materialamount";
+            public static final String MATERIAL_REQUIRED = "schematica.gui.materialrequired";
```
Comment (Kamesuta) [inline src/main/java/com/github/lunatrius/schematica/reference/Names.java]: Define for localization

File: src/main/resources/assets/schematica/lang/en_US.lang [hunk 11]
```diff
@@ -28,6 +28,8 @@ schematica.gui.z=Z:
 schematica.gui.materials=Materials
 schematica.gui.materialname=Material
 schematica.gui.materialamount=Amount
+schematica.gui.materialrequired=Required
```
Comment (Kamesuta) [inline src/main/resources/assets/schematica/lang/en_US.lang]: English Localization

File: src/main/resources/assets/schematica/lang/ja_JP.lang [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+schematica.gui.saveselection=選択範囲の保存
+schematica.gui.on=有効
+schematica.gui.off=無効
+schematica.gui.x=X:
+schematica.gui.y=Y:
+schematica.gui.z=Z:
+schematica.gui.materials=建材
+schematica.gui.materialname=建材
+schematica.gui.materialamount=量
+schematica.gui.materialrequired=必要数
```
Comment (Kamesuta) [inline src/main/resources/assets/schematica/lang/ja_JP.lang]: Japanese Localization

**Comment by Lunatrius** [comment]
Looks great! I have one final request. I'll be most likely merging this into master (1.9) and it might get annoying try to merge 11 commits separately due to some differences in the code -- could you squash them?

For the command line:
```bash
git rebase -i HEAD~11

--- # replace 'pick' with 's' on lines 2-11 (this will open in an editor)
pick 73462a1 Require Item Counts
s cdba202 Scanning Inventory and Count
s e357d6d Display require item count
s 120d7d7 Stack Display seems to be useful :-)
s c908fbd Solve a issue, "The import reordering is just adding noise"
s f6fd3d4 Solve a issue, "there should already be a method in LunatriusCore"
s 0cff0ed Solve item stack sizes and Hide it when amount is less than stack size.
s 08bb3c6 Remove noisy one line comments
s dbf97cb Always display "Available" when player is creative mode.
s 6e3139c Fix item damage comparing
s 3060399 LOCALIZATION

--- # write the message for all commits
--- # push the changes to GitHub
git push --force
```

TortoiseGit: http://stackoverflow.com/a/21306306

**Comment by Kamesuta** [comment]
I'm glad you to say such a thing. :-)
**Squash** has just **completed**.
(I renamed a commit message to your **great** one.)
Good luck :)

**Comment by Lunatrius** [comment]
Thanks! I'll try to get a build out tomorrow and port over the changes to 1.9 as well.

**Pull request merged**